plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Internal.hs

```diff
@@ -101,27 +106,28 @@ getSemanticTokensRule recorder =
     ast <- handleMaybe (LogNoAST $ show nfp) $ getAsts hieAst Map.!? (HiePath . mkFastString . fromNormalizedFilePath) nfp
     virtualFile <- handleMaybeM LogNoVF $ getVirtualFile nfp
     -- get current location from the old ones
-    let spanNamesMap = hieAstSpanNames virtualFile ast
-    let names = nameSetElemsStable $ unionNameSets $ Map.elems spanNamesMap
-    let localSemanticMap = mkLocalNameSemanticFromAst names (hieKindFunMasksKind hieKind) refMap
+    let spanIdMap = M.filter (not . null) $ hieAstSpanIdentifiers virtualFile ast
+    let names = S.toList $ S.unions $ Map.elems spanIdMap
+    let localSemanticMap = mkLocalIdSemanticFromAst names (hieKindFunMasksKind hieKind) refMap
     -- get imported name semantic map
-    let importedNameSemanticMap = foldr (getTypeExclude localSemanticMap getTyThingMap) emptyNameEnv names
-    let sMap = plusNameEnv_C (<>) importedNameSemanticMap localSemanticMap
-    let rangeTokenType = extractSemanticTokensFromNames sMap spanNamesMap
+    let importedIdSemanticMap = foldr (getTypeExclude localSemanticMap getTyThingMap) mempty names
```

Comment:
good idea, done

Reply:
Did you still mean to use `foldr` here? Why not `fromSet`?

Reply:
I was under the impression using `fromSet` would still have the same time complexity since I was using `elem` in getTypeExclude.

Now, giving a second thought, combining `fromSet` with `Set.difference` does reduce time complexity a bit for some cases.  reviced!